src/Functions/toStartOfInterval.cpp

```diff
@@ -129,18 +185,24 @@ class FunctionToStartOfInterval : public IFunction
             case ResultType::Date:
                 return std::make_shared<DataTypeDate>();
             case ResultType::DateTime:
-                return std::make_shared<DataTypeDateTime>(extractTimeZoneNameFromFunctionArguments(arguments, 2, 0, false));
+            {
+                const size_t time_zone_arg_num = (overload == Overload::Default) ? 2 : 3;
+                return std::make_shared<DataTypeDateTime>(extractTimeZoneNameFromFunctionArguments(arguments, time_zone_arg_num, 0, false));
+            }
             case ResultType::DateTime64:
             {
                 UInt32 scale = 0;
+                if (isDateTime64(arguments[0].type))
+                    scale = assert_cast<const DataTypeDateTime64 &>(*arguments[0].type.get()).getScale();
                 if (interval_type->getKind() == IntervalKind::Kind::Nanosecond)
-                    scale = 9;
+                    scale = (9 > scale) ? 9 : scale;
                 else if (interval_type->getKind() == IntervalKind::Kind::Microsecond)
-                    scale = 6;
+                    scale = (6 > scale) ? 6 : scale;
                 else if (interval_type->getKind() == IntervalKind::Kind::Millisecond)
-                    scale = 3;
+                    scale = (3 > scale) ? 3 : scale;
```

Comment:
This part doesn't really make sense because 
- we don't support milli- micro- nanoseconds with non-DateTime64 values
- with DateTime64 values and scale >= second we use the standard scale

Reply:
Actually, it changes the precision on bigger in case of bigger scale of datetime argument, but I think we don't need it